### PR TITLE
Auto (with codegen) binary serializer 

### PR DESCRIPTION
**Pull request by NikVolf**
codegen for serialization so that

```
#[derive(Binary)]
pub enum Root {
	Top,
	Middle(u32, u64),
}
```

could be serialized without additional code via

```
let res = ::ipc::binary::serialize_into(&Root::Middle(10, 100), &mut socket);
```

- with the exception of header info, no intermediate allocations, whole payloads serializes from the one stream of bytes
- packed (size 0) representation of some values (Option::None, empty vector, etc..)
- ... a lot of space for optimizations

### CODE REVIEW

File: ipc/codegen/src/serialization.rs [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+		Annotatable::Item(ref item) => item,
+		_ => {
+			cx.span_err(meta_item.span, "`#[derive(Binary)]` may only be applied to structs and enums");
+			return;
+		}
+	};
+
+	let builder = aster::AstBuilder::new().span(span);
+
+    let impl_item = match serialize_item(cx, &builder, &item) {
```
Comment (gavofyork) [inline ipc/codegen/src/serialization.rs]: spaceS!!!!!

File: ipc/codegen/src/serialization.rs [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+    ).unwrap())
+}
+
+#[allow(unreachable_code)]
+fn binary_expr(
+    cx: &ExtCtxt,
+    builder: &aster::AstBuilder,
+    item: &Item,
+    impl_generics: &ast::Generics,
+    ty: P<ast::Ty>,
```
Comment (gavofyork) [inline ipc/codegen/src/serialization.rs]: spaces!!!

**Comment by gavofyork** [comment]
files have polluted whitespace

**Comment by NikVolf** [comment]
fixed :)

File: ipc/codegen/src/serialization.rs [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+
+		let member_expr = match value_ident {
+			Some(x) => {
+				let field_id = builder.id(field.ident.unwrap());
+				quote_expr!(cx, $x . $field_id)
+			},
+			None => {
+				let index_ident = builder.id(format!("__field{}", index));
+				quote_expr!(cx, $index_ident)
+			}
```
Comment (gavofyork) [inline ipc/codegen/src/serialization.rs]: either don't include the comma after the previous arm or put one here

File: ipc/codegen/src/serialization.rs [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+		false => {
+			if value_ident.is_some() {
+				let instance_create = named_fields_sequence(cx, &ty, fields);
+				quote_expr!(cx, { $map_stmts; $instance_create; Ok(result) })
+			}
+			else {
+				let map_variant = P(fields_sequence(cx, &ty, fields, &instance_ident.unwrap_or(builder.id("Self"))));
+				quote_expr!(cx, { $map_stmts; Ok($map_variant) })
+			}
+		}
```
Comment (gavofyork) [inline ipc/codegen/src/serialization.rs]: same

File: ipc/codegen/src/serialization.rs [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+		ast::VariantData::Tuple(ref fields, _) => {
+			binary_expr_struct(
+				cx,
+				&builder,
+				ty,
+				fields,
+				Some(builder.id("self")),
+				None,
+			)
+		}
```
Comment (gavofyork) [inline ipc/codegen/src/serialization.rs]: same

**Comment by gavofyork** [comment]
three reasonable stylistic forms of `match`:

one-liner arms end in `,`, apart from last entry:
```
match a { 0 => x, 1 => y, 2 => z }
```

for multi-liners either all arms end in `,` or (if they're all `{...}` bodies) none do:
```
match a {
  0 => {x},
  1 => {y},
  2 => {z},
}
```

or
```
match a {
  0 => {x}
  1 => {y}
  2 => {z}
}
```

happy to discuss alterations to these conventions, but that's what i had in my mind so far.

**Pull request merged**